pallets/aggregate/src/lib.rs

```diff
@@ -372,21 +391,39 @@ pub mod pallet {
                     next_aggregation_id,
                     max_aggregation_size,
                     publish_queue_size,
+                    aggregate_rules,
                     ticket,
+                    delivery,
                 )))
             }
         }
 
         /// Compute and reserve the currency for further publication
-        fn reserve_currency_for_publication(
+        fn reserve_currency(
             &self,
             account: &AccountOf<T>,
-        ) -> Result<BalanceOf<T>, DispatchError> {
+        ) -> Result<Reserve<BalanceOf<T>>, DispatchError> {
             let estimated = estimate_publish_aggregation_fee::<T>(self.max_aggregation_size);
-            let hold = (estimated.defensive_saturating_add(
+            let aggregate = (estimated.defensive_saturating_add(
                 <T as Config>::ComputePublisherTip::compute_tip(estimated).unwrap_or_default(),
             )) / self.next.size.into();
-            T::Hold::hold(&HoldReason::Aggregation.into(), account, hold).map(|_| hold)
+            let delivery = *self.delivery.price() / self.next.size.into();
+
+            T::Hold::hold(&HoldReason::Aggregation.into(), account, aggregate)?;
+            T::Hold::hold(&HoldReason::Delivery.into(), account, delivery).inspect_err(|_| {
+                T::Hold::release(
```

Comment:
Why the release is needed ? If the second hold fails, the whole extrinsic fails, thus the state stays untouched no ?

Reply:
If reserve currency fails the `on_proof_verified()` call *dosen't fail* but just emit an event. Anyway, I wrote a test about it (before writing this code) and if you remove this release call you'll see the test to fail.

If we don't release it in the case where you have enough money for aggregate, but not for delivery your money held for aggregation will remain on held forever.